packages/@lwc/ssr-compiler/src/transmogrify.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { traverse, builders as b } from 'estree-toolkit';
+import { produce } from 'immer';
+import type { Node } from 'estree';
+import type { Program as EsProgram } from 'estree';
+
+export type TransmogrificationMode = 'sync' | 'async';
+
+interface TransmogrificationState {
+    mode: TransmogrificationMode;
+}
+
+export type Visitors = Parameters<typeof traverse<Node, TransmogrificationState>>[1];
+
+const EMIT_IDENT = b.identifier('$$emit');
+
+const visitors: Visitors = {
+    $: { scope: true },
+    FunctionDeclaration(path, state) {
+        const { node } = path;
+        if (!node?.async || !node?.generator) {
+            return;
+        }
+        node.generator = false;
+        node.async = state.mode === 'async';
+        node.params.unshift(EMIT_IDENT);
+    },
+    YieldExpression(path, state) {
+        const { node } = path;
+        if (!node) {
+            return;
+        }
+
+        if (node.delegate) {
+            // transform `yield* foo(arg)` into `foo($$emit, arg)` or `await foo($$emit, arg)`
+            if (node.argument?.type !== 'CallExpression') {
+                throw new Error(
+                    'Implementation error: cannot transmogrify complex yield-from expressions'
+                );
+            }
```

Comment:
This is another example of where we may want additional protections. If the component author utilizes `yield` expressions in their own code, that might result in a compilation error here.

There are probably a small set of function-names that we could check for here and, if the yield-expression's argument is not a `CallExpression` or if the function being called does not conform to the set of function-names we're looking for, perhaps we just return early here.

Reply:
This was also addressed with new compile-time protections.

Reply:
Let's add a comment explaining that this is impossible now. Ditto for above.